src/java/org/apache/cassandra/config/DataStorageSpec.java

```diff
@@ -255,6 +219,210 @@ public String toString()
+        {
+            super(quantity, unit, BYTES, Integer.MAX_VALUE);
+        }
+
+        /**
+         * Creates a {@code SmallestDataStorage.IntBytesBound} of the specified amount in bytes.
+         *
+         * @param bytes where bytes shouldn't be bigger than Integer.MAX_VALUE-1
+         */
+        public IntBytesBound(long bytes)
+        {
+            this(bytes, BYTES);
+        }
+    }
+
+    /**
+     * Represents a data storage used for Cassandra configuration. The bound is [0; Integer.MAX_VALUE) in kibibytes.
+     * If the user sets a different unit - we still validate that converted to kibibytes the quantity will not exceed
+     * that upper bound. (CASSANDRA-17571)
+     */
+    public final static class IntKibibytesBound extends DataStorageSpec
+    {
+        /**
+         * Creates a {@code DataStorageSpec.IntKibibytesBound} of the specified amount.
+         *
+         * @param value the data storage
+         *
+         */
+        public IntKibibytesBound(String value)
+        {
+            super(value, KIBIBYTES, Integer.MAX_VALUE);
+        }
+
+        /**
+         * Creates a {@code DataStorageSpec.IntKibibytesBound} of the specified amount in the specified unit.
+         *
+         * @param quantity where quantity shouldn't be bigger than Integer.MAX_VALUE - 1 in kibibytes
+         * @param unit in which the provided quantity is
+         */
+        public IntKibibytesBound(long quantity, DataStorageUnit unit)
+        {
+            super(quantity, unit, KIBIBYTES, Integer.MAX_VALUE);
+        }
+
+        /**
+         * Creates a {@code DataStorageSpec.IntKibibytesBound} of the specified amount in kibibytes.
+         *
+         * @param kibibytes where kibibytes shouldn't be bigger than Integer.MAX_VALUE-1
+         */
+        public IntKibibytesBound(long kibibytes)
+        {
+            this(kibibytes, KIBIBYTES);
+        }
+    }
+
+    /**
+     * Represents a data storage used for Cassandra configuration. The bound is [0; Long.MAX_VALUE) in mebibytes.
+     * If the user sets a different unit - we still validate that converted to mebibytes the quantity will not exceed
+     * that upper bound. (CASSANDRA-17571)
+     */
+    public final static class LongMebibytesBound extends DataStorageSpec
+    {
+        /**
+         * Creates a {@code DataStorageSpec.LongMebibytesBound} of the specified amount.
+         *
+         * @param value the data storage
+         */
+        public LongMebibytesBound(String value)
+        {
+            super(value, MEBIBYTES, Long.MAX_VALUE);
+        }
+
+        /**
+         * Creates a {@code DataStorageSpec.LongMebibytesBound} of the specified amount in the specified unit.
+         *
+         * BE CAREFUL, IF YOU DECIDE TO USE UNIT BYTES OR KIBIBYTES, SET A NUMBER THAT WILL NOT LEAD TO LOSS OF PRECISION DURING CONVERSION
+         * TO MEBIBYTES. WE GUARD FOR THIS IN THE PREVIOUS CONSTRUCTOR BUT NOT THIS ONE
```

Comment:
I think we would need similar warnings in the two classes for kibibytes, since those could also lose precision if bytes are used. Also, we could specify what type of precision loss is going to be, is it going to be rounded up or down? I wonder if there is any case where that precision loss is useful or we should just offer the same validation in both constructors.

Reply:
`I think we would need similar warnings in the two classes for kibibytes`
Yes, I missed to add it :-( I will take care, thanks
About the precision, well 0.4 and 0.6 will be 0

What bothers me is the  bigger units we support. 3.9 days will be considered 3 days. I would prefer not to go down that road. 
Now the only reason I left these constructors without validation is one special case we have and that those constructors will be used carefully only by us, the C* developers. Thinking more, maybe I should leave without validation only that constructor so it is more like edge case supported than general support and people can think twice before doing something similar. Probably it will be safer not to encourage it. 

The case I talk about is `max_value_size` - `DatabaseDescriptor#setMaxValueSize `

Reply:
> 3.9 days will be considered 3 days

Still just looking through this first batch of comments and not the patch itself, but if there's any reasonable way to make specifying fractional days illegal, I would do it.

Reply:
Added validation to all constructors except that special case one where I left a explicit comment and instructions